### PR TITLE
CASSSIDECAR-173: Endpoint to retrieve C* gossip status

### PR DESCRIPTION
**Pull request by skoppu22**
https://issues.apache.org/jira/browse/CASSSIDECAR-173
Developing a new Sidecar endpoint to retrieve gossip running status of Cassandra, so the end user can use this endpoint without having to invoke C* JMX call. This endpoint will have authorization checks enabled once authorization feature is added to the Sidecar.

Circle CI run: https://app.circleci.com/pipelines/github/skoppu22/cassandra-sidecar?branch=gossips2

### CODE REVIEW

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        this.gossipRunning = gossipRunning;
+    }
+
+    @JsonProperty("gossipRunning")
+    public boolean gossipRunning()
+    {
+        return gossipRunning;
+    }
+
+    private final boolean gossipRunning;
```
Comment (bbotella) [inline client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Define this at the beginning of the class

File: server/src/main/java/org/apache/cassandra/sidecar/server/MainModule.java [hunk 2]
```diff
@@ -380,6 +382,9 @@ public Router vertxRouter(Vertx vertx,
         router.get(ApiEndpointsV1.GOSSIP_INFO_ROUTE)
               .handler(gossipInfoHandler);
 
+        router.get(ApiEndpointsV1.GOSSIP_STATUS_ROUTE)
```
Comment (bbotella) [inline server/src/main/java/org/apache/cassandra/sidecar/server/MainModule.java]: It is really unfortunate having to add an entire new API when we already have the GOSSIP_INFO_ROUTE endpoint? I even see that the GossipInfoResponse already has a deprecated STATUS field. 

I understand that we are trying to make this call more efficient by only checking the gossip status. I wonder, could we add a filtering parameter that helps us only executing the part of the handler that we want?

I'm thinking of a filtering pattern like:
`/api/v1/cassandra/gossip?filter=['status']` or something along those lines?

That way, and finding out why/where it was deprecated, maybe we can hijack it to use this `storageOperations::isGossipRunning` can save us from having to add this new API

Comment (skoppu22) [reply server/src/main/java/org/apache/cassandra/sidecar/server/MainModule.java]: ?filter=<> doesn't sound right in this case, as filter is usually used to chose some/partial fields of the output which was returned without using filter. In this case /gossip endpoint doesn't return 'status', so we can't `filter` status from it. Let me see what are alternatives and will get back to you.

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 1 above)
Comment (skoppu22) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Most of Sidecar request/response classes following this style. I am also used to the style of defining at the beginning of the class, but just trying to be inline with existing classes. May be we can start changing the style now, as easy to find all variables right at the beginning of the class.

Comment (frankgh) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: > Most of Sidecar request/response classes following this style

This is not true, I'm not sure which class you are looking at, but all classes under `client-common/src/main/java/org/apache/cassandra/sidecar/common/response` define fields at the beginning of the class. And we should follow the Cassandra code structure https://cassandra.apache.org/_/development/code_style.html under Code Structure section > Class Layout.

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public GossipStatusResponse(@JsonProperty("gossipRunning") boolean gossipRunning)
+    {
+        this.gossipRunning = gossipRunning;
+    }
+
+    @JsonProperty("gossipRunning")
+    public boolean gossipRunning()
+    {
+        return gossipRunning;
```
Comment (frankgh) [inline client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: returning a boolean becomes inflexible if you need to support a use case where you have more than one value. I suggest we change this to a payload that looks like 

```json
{
   "status": "RUNNING"
}
```

and 

```json
{
   "status": "NOT_RUNNING"
}
```

File: client/src/main/java/org/apache/cassandra/sidecar/client/SidecarClient.java [hunk 4]
```diff
@@ -215,6 +216,16 @@ public CompletableFuture<GossipInfoResponse> gossipInfo()
...
     }
 
+    /**
+     * Executes the GET gossip status request using the default retry policy and configured selection policy
+     *
+     * @return a completable future with gossip status response
+     */
+    public CompletableFuture<GossipStatusResponse> gossipStatus()
+    {
+        return executor.executeRequestAsync(requestBuilder().gossipStatusRequest().build());
```
Comment (frankgh) [inline client/src/main/java/org/apache/cassandra/sidecar/client/SidecarClient.java]: this request should be instance-specific

File: server-common/src/main/java/org/apache/cassandra/sidecar/common/server/StorageOperations.java [hunk 5]
```diff
@@ -102,4 +103,9 @@ default void outOfRangeDataCleanup(@NotNull String keyspace, @NotNull String tab
     {
         outOfRangeDataCleanup(keyspace, table, 1);
     }
+
+    /**
+     * @return returns gossip status
+     */
+    GossipStatusResponse isGossipRunning();
```
Comment (frankgh) [inline server-common/src/main/java/org/apache/cassandra/sidecar/common/server/StorageOperations.java]: Maybe better aligned with the action we are trying to perform
```suggestion
    GossipStatusResponse gossipStatus();
```

File: server/src/main/java/org/apache/cassandra/sidecar/routes/AbstractHandler.java [hunk 6]
```diff
@@ -331,4 +332,16 @@ public static String extractHostAddressWithoutPort(String address)
...
+    {
+        CassandraAdapterDelegate delegate = this.metadataFetcher.delegate(host);
+        StorageOperations storageOperations = delegate == null ? null : delegate.storageOperations();
+        if (storageOperations == null)
+        {
+            throw cassandraServiceUnavailable();
+        }
+
+        return storageOperations;
+    }
```
Comment (frankgh) [inline server/src/main/java/org/apache/cassandra/sidecar/routes/AbstractHandler.java]: we already have a method that does something similar in this class, we should just generalize that method. I suggest renaming `ifMetricsOpsAvailable` , and the code would look like this instead

```
    protected <P> void ifAvailableFromDelegate(RoutingContext context,
                                               String host,
                                               Function<CassandraAdapterDelegate, P> mapper,
                                               Consumer<P> ifAvailable)
    {
        CassandraAdapterDelegate delegate = metadataFetcher.delegate(host);
        if (delegate == null)
        {
            context.fail(cassandraServiceUnavailable());
            return;
        }
        P applied = mapper.apply(delegate);
        if (applied == null)
        {
            context.fail(cassandraServiceUnavailable());
            return;
        }
        ifAvailable.accept(applied);
    }
```

The existing callsite would change to this: 

```
ifAvailableFromDelegate(context, host, CassandraAdapterDelegate::metricsOperations, operations -> {
```

File: server/src/main/java/org/apache/cassandra/sidecar/routes/GossipStatusHandler.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    @Override
+    public void handleInternal(RoutingContext context,
+                               HttpServerRequest httpRequest,
+                               String host,
+                               SocketAddress remoteAddress,
+                               Void request)
+    {
+        StorageOperations storageOperations = getStorageOperations(host);
+        logger.debug("Retrieving gossip status, remoteAddress={}, instance={}",
```
Comment (frankgh) [inline server/src/main/java/org/apache/cassandra/sidecar/routes/GossipStatusHandler.java]: this log entry is redundant, we already log 

```
logger.debug("{} received request={}, remoteAddress={}, instance={}",
                         this.getClass().getSimpleName(), requestParams, remoteAddress, host);
```

for all requests in handlers of the `AbstractHandler` type.

File: server/src/test/integration/org/apache/cassandra/sidecar/routes/GossipStatusHandlerIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Test GET gossip status endpoint with C* container
+ */
+@ExtendWith(VertxExtension.class)
+public class GossipStatusHandlerIntegrationTest extends IntegrationTestBase
+{
+    private static final String testRoute = "/api/v1/cassandra/gossip/status";
+
+    @CassandraIntegrationTest()
+    void testGossipDisabled(CassandraTestContext context, VertxTestContext testContext)
```
Comment (frankgh) [inline server/src/test/integration/org/apache/cassandra/sidecar/routes/GossipStatusHandlerIntegrationTest.java]: can we combine both test cases into a single test case? these are expensive to spin up, so whenever possible we should reuse the in-jvm dtest apparatus available .

File: server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Common test code to setup/teardown the server, populate test instances metadata and mocks for JMX calls
+ */
+class JmxCommonTest
+{
+    static final Logger LOGGER = LoggerFactory.getLogger(JmxCommonTest.class);
```
Comment (frankgh) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java]: if we already have an in-jvm dtest this becomes less useful IMO. There is no need to do mocking if we have the actual server running

**Review by frankgh — COMMENTED** [review]
Added some comments to the PR

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 1 above)
Comment (skoppu22) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: This one I followed https://github.com/apache/cassandra-sidecar/blob/trunk/client-common/src/main/java/org/apache/cassandra/sidecar/common/response/ConnectedClientStatsResponse.java

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 3 above)
Comment (skoppu22) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Client need to rely on deterministic values to parse the response and take actions accordingly. String "RUNNING" or "NOT_RUNNING" can break the client even with any minor changes or extra spaces etc in the output. Instead, if it is possible to return an enum as rest API response, we can use an enum. Otherwise boolean value true/false would be deterministic for the caller to parse the output

File: server/src/main/java/org/apache/cassandra/sidecar/server/MainModule.java (hunk 2 above)
Comment (skoppu22) [reply server/src/main/java/org/apache/cassandra/sidecar/server/MainModule.java]: As we discussed offline, we cannot infer gossip running status from STATUS (or STATUS_WITH_PORT) fields of gossip info. Hence we cannot use gossip/filter=status for gossip running status, as that filter can be mistaken for STATUS field of gossip info. Will continue with gossip/status endpoint (Other option could be gossip/runningStatus, but sounds too verbose)

File: server/src/main/java/org/apache/cassandra/sidecar/routes/AbstractHandler.java (hunk 6 above)
Comment (skoppu22) [reply server/src/main/java/org/apache/cassandra/sidecar/routes/AbstractHandler.java]: Nice

File: server/src/test/integration/org/apache/cassandra/sidecar/routes/GossipStatusHandlerIntegrationTest.java (hunk 8 above)
Comment (skoppu22) [reply server/src/test/integration/org/apache/cassandra/sidecar/routes/GossipStatusHandlerIntegrationTest.java]: OK

File: server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java (hunk 9 above)
Comment (skoppu22) [reply server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java]: - We do more test scenarios using mock, especially related to failure cases as mock makes it easy to throw exceptions or return invalid values
- Can we get code coverage measured with in-jvm dtests? For example, someone using open source Sidecar and hooking to their build systems, can they get code coverage numbers with in-jvm dtests?

File: client/src/main/java/org/apache/cassandra/sidecar/client/SidecarClient.java (hunk 4 above)
Comment (skoppu22) [reply client/src/main/java/org/apache/cassandra/sidecar/client/SidecarClient.java]: Thanks for noticing this

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 3 above)
Comment (bbotella) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Well, both enum or boolean flags are technically strings that need to be parsed/interpreted by the client as well. I do like the idea @frankgh  is proposing of having the enum with two possible states, that allows us to easily expand the amount of information we can provide.

Comment (skoppu22) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Why boolean is treated as a string? Client can prepare GossipStatusResponse from the json received in the response and can get boolean using GossipStatusResponse.gossipRunning()

Comment (bbotella) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: I meant that a response payload is a bunch of bytes that need to be parsed anyway, even if it is a json payload, it needs to be parsed. The client is in charge of parsing it and interpreting fields as booleans, strings, or anything else.

That said, I do agree with @frankgh in which the flexibility added by returning that enum is preferable in this case. On the client side, it is client's responsibility to parse the response correctly. In our current implementation, we control de java client for instance. We are also adding OpenAPI documentation that will help clients parse the given responses as part of this [PR](https://github.com/apache/cassandra-sidecar/pull/164). 

Long story short, I don't think returning a `"RUNNING"` response is making things harder for the client than returning a `true`, and it gives us flexibility to grow the API in the future.

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 1 above)
Comment (frankgh) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: This is something we should fix

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 3 above)
Comment (skoppu22) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Updated to String

File: server/src/main/java/org/apache/cassandra/sidecar/routes/AbstractHandler.java (hunk 6 above)
Comment (skoppu22) [reply server/src/main/java/org/apache/cassandra/sidecar/routes/AbstractHandler.java]: Looks like this is no longer needed as you moved null checks into getters

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class GossipStatusResponse
+{
+    private final String gossipRunningStatus;
+
+    /**
+     * Constructs a response with gossip status
+     * @param gossipRunningStatus running status of gossip
+     */
+    public GossipStatusResponse(@JsonProperty("gossipRunningStatus") String gossipRunningStatus)
```
Comment (frankgh) [inline client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: I think `status` should be sufficient. We don't have a precedent for qualifying the field name like this.

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        this.gossipRunningStatus = gossipRunningStatus;
+    }
+
+    @JsonProperty("gossipRunningStatus")
+    public String gossipRunningStatus()
```
Comment (frankgh) [inline client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: should we add a utility method here `isRunning` that returns `"RUNNING".equals(gossipRunningStatus)`. Similar to what you had in an earlier iteration of this PR.

File: server-common/src/main/java/org/apache/cassandra/sidecar/common/server/StorageOperations.java [hunk 12]
```diff
@@ -102,4 +103,9 @@ default void outOfRangeDataCleanup(@NotNull String keyspace, @NotNull String tab
     {
         outOfRangeDataCleanup(keyspace, table, 1);
     }
+
+    /**
+     * @return returns gossip status
```
Comment (frankgh) [inline server-common/src/main/java/org/apache/cassandra/sidecar/common/server/StorageOperations.java]:
```suggestion
     * @return the gossip status
```

File: server/src/main/java/org/apache/cassandra/sidecar/routes/GossipStatusHandler.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.net.SocketAddress;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.cassandra.sidecar.common.server.StorageOperations;
+import org.apache.cassandra.sidecar.concurrent.ExecutorPools;
+import org.apache.cassandra.sidecar.utils.InstanceMetadataFetcher;
+
+
+/**
+ * Handler for to retrieve gossip status
```
Comment (frankgh) [inline server/src/main/java/org/apache/cassandra/sidecar/routes/GossipStatusHandler.java]:
```suggestion
 * Handles gossip status retrieval
```

File: server/src/test/integration/org/apache/cassandra/sidecar/routes/GossipStatusHandlerIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                     .expect(ResponsePredicate.SC_OK)
+                     .send()
+                     .compose(response -> {
+                         assertThat(response.statusCode()).isEqualTo(OK.code());
+                         return Future.succeededFuture(response.bodyAsJson(GossipStatusResponse.class));
+                     });
+    }
+
+    @CassandraIntegrationTest()
+    void testGossipStatus(CassandraTestContext context, VertxTestContext testContext)
```
Comment (frankgh) [inline server/src/test/integration/org/apache/cassandra/sidecar/routes/GossipStatusHandlerIntegrationTest.java]: nice!

File: server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * Tests for {@link GossipInfoHandler}
+ */
+@ExtendWith(VertxExtension.class)
+public class GossipStatusHandlerTest extends JmxCommonTest
+{
+    private static final String testRoute = "/api/v1/cassandra/gossip/status";
+
+    @Test
+    void testGossipRunning(VertxTestContext context)
```
Comment (frankgh) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java]: this test is redundant, not sure what we gain here.

File: server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        when(storageOperations.gossipStatus()).thenReturn(new GossipStatusResponse("RUNNING"));
+
+        WebClient client = WebClient.create(vertx);
+        client.get(server.actualPort(), "127.0.0.1", testRoute)
+              .expect(ResponsePredicate.SC_OK)
+              .send(context.succeeding(response -> verifyValidResponse(context, response, "RUNNING")));
+    }
+
+    @Test
+    void testGossipNotRunning(VertxTestContext context)
```
Comment (frankgh) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java]: this is also redundant not sure what exactly we are testing here.

File: server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        when(storageOperations.gossipStatus()).thenReturn(new GossipStatusResponse("NOT_RUNNING"));
+
+        WebClient client = WebClient.create(vertx);
+        client.get(server.actualPort(), "127.0.0.1", testRoute)
+              .expect(ResponsePredicate.SC_OK)
+              .send(context.succeeding(response -> verifyValidResponse(context, response, "NOT_RUNNING")));
+    }
+
+    @Test
+    void testWithInstanceId(VertxTestContext context)
```
Comment (frankgh) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java]: also, this test doesn't make a lot of sense to me. My problem with the test is that in the setup `JmxCommonTest` you share the `delegate` and `storageOperations` among 3 instances. This setup is problematic because in the real execution of sidecar each instance will have its own `delegate` and all the operations associated with a single delegate. I would drop this test, we already cover accessing different instances with the query parameter, so I would consider testing this outside of the scope of this test.

File: server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        WebClient client = WebClient.create(vertx);
+        client.get(server.actualPort(), "127.0.0.1", testRoute)
+              .expect(ResponsePredicate.SC_INTERNAL_SERVER_ERROR)
+              .send(context.succeeding(response -> {
+                  assertThat(response.statusCode()).isEqualTo(INTERNAL_SERVER_ERROR.code());
+                  context.completeNow();
+              }));
+    }
+
+    private void verifyValidResponse(VertxTestContext context, HttpResponse<Buffer> response, String expectedValue)
```
Comment (frankgh) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java]: maybe call this `verifyAndComplete`

File: server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+            LOGGER.error("Close event timed out.");
+    }
+
+    class JmxTestModule extends AbstractModule
+    {
+        @Provides
+        @Singleton
+        public InstancesMetadata instanceConfig()
+        {
+            int instanceId1 = 100;
```
Comment (frankgh) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java]: I think we should probably model this correctly if we are meaning to use this in other tests.

**Review by frankgh — COMMENTED** [review]
Looking good, I have a few additional comments.

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 1 above)
Comment (yifan-c) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Can it be enforced with code style? We want to layout the class in this order. 
```
// variables 
public static 
protected static
private static
public 
protected
private 
// methods 
// same ordering goes again
// visible for testing methods
// then, inner class
```

**Review by yifan-c — CHANGES_REQUESTED** [review]
Thanks for the contribution.

First of all, please add an entry to `CHANGES.txt`.

Second, the patch is to add an API to check whether gossip is enabled or disabled in the companion Cassandra node. There are existing health APIs that check the status of corresponding components. See the following code block for example. 

```
GET /api/v1/cassandra/native/__health
{"status" : "OK | NOT_OK"}

GET /api/v1/cassandra/jmx/__health
{"status" : "OK | NOT_OK"}
```

I think we should be consistent with the existing pattern. In other words, the API should be this.

```
GET /api/v1/cassandra/gossip/__health
{"status" : "OK | NOT_OK"}
```

File: adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraStorageOperations.java [hunk 20]
```diff
@@ -213,4 +214,11 @@ public void outOfRangeDataCleanup(@NotNull String keyspace, @NotNull String tabl
         jmxClient.proxy(StorageJmxOperations.class, STORAGE_SERVICE_OBJ_NAME)
                  .forceKeyspaceCleanup(concurrency, keyspace, table);
     }
+
+    @Override
+    public GossipStatusResponse gossipStatus()
+    {
+        return new GossipStatusResponse(jmxClient.proxy(StorageJmxOperations.class, STORAGE_SERVICE_OBJ_NAME)
+                                                 .isGossipRunning() ? "RUNNING" : "NOT_RUNNING");
```
Comment (bbotella) [inline adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraStorageOperations.java]: What do you think about moving `"RUNNING"` and `"NOT_RUNNING"` to an enum?

**Comment by frankgh** [comment]
> I think we should be consistent with the existing pattern. In other words, the API should be this.

I agree with Yifan, we should try to stick closer to the existing patterns. +1 on this:

```
GET /api/v1/cassandra/gossip/__health
{"status" : "OK | NOT_OK"}
```

File: adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraStorageOperations.java (hunk 20 above)
Comment (bbotella) [reply adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraStorageOperations.java]: I actually agree with @yifan-c  suggestion. So, what about having the enum for the `"OK"` and `"NOT_OK"`?

File: client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java (hunk 10 above)
Comment (skoppu22) [reply client-common/src/main/java/org/apache/cassandra/sidecar/common/response/GossipStatusResponse.java]: Just calling it 'status' can be mistaken with STATUS field of gossip info, as we discussed before regarding this field. Gossip running status just tells is gossip thread running in C*. Where as STATUS field has different meaning. So we need to be clear that this is gossip thread running status

**Comment by skoppu22** [comment]
> Thanks for the contribution.
> 
> First of all, please add an entry to `CHANGES.txt`.
> 
> Second, the patch is to add an API to check whether gossip is enabled or disabled in the companion Cassandra node. There are existing health APIs that check the status of corresponding components. See the following code block for example.
> 
> ```
> GET /api/v1/cassandra/native/__health
> {"status" : "OK | NOT_OK"}
> 
> GET /api/v1/cassandra/jmx/__health
> {"status" : "OK | NOT_OK"}
> ```
> 
> I think we should be consistent with the existing pattern. In other words, the API should be this.
> 
> ```
> GET /api/v1/cassandra/gossip/__health
> {"status" : "OK | NOT_OK"}
> ```


This API is returning C* gossip thread running status (just tells is gossip thread running or not). Having thread running doesn't mean gossip health is OK. So I wouldn't prefer to call it 'gossip health'. Gossip thread may be running, but gossip state might be not healthy at C* side

File: server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java (hunk 15 above)
Comment (skoppu22) [reply server/src/test/java/org/apache/cassandra/sidecar/routes/GossipStatusHandlerTest.java]: The gain depends on answer to https://github.com/apache/cassandra-sidecar/pull/159/files#r1886797448

Do we get code coverage with in-jvm tests? If yes, we can remove this test

File: server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java (hunk 19 above)
Comment (skoppu22) [reply server/src/test/java/org/apache/cassandra/sidecar/routes/JmxCommonTest.java]: We can remove multi instance stuff from this. But I see almost all tests in this directory are having module extending AbstractModule and using mocks, I believe we can move common code to a common test framework file and reuse in most of tests

File: server/src/test/java/org/apache/cassandra/sidecar/routes/CommonTest.java [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.cassandra.sidecar.server.MainModule;
+import org.apache.cassandra.sidecar.server.Server;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Common test code to setup/teardown the server and populate test instances metadata
+ */
+public class CommonTest
```
Comment (skoppu22) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/CommonTest.java:52]: This class contains common test setup and teardown which repeated in almost all test files. For now I am using this common in GossipInfoHandlerTest and GossipHealthHandlerTest. Will open a separate PR to modify other test files to use this common test class wherever possible

File: server/src/main/java/org/apache/cassandra/sidecar/routes/GossipHealthHandler.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+                               HttpServerRequest httpRequest,
+                               String host,
+                               SocketAddress remoteAddress,
+                               Void request)
+    {
+        StorageOperations operations = metadataFetcher.delegate(host).storageOperations();
+        executorPools.service()
+                         .executeBlocking(operations::isGossipRunning)
+                         .onSuccess(isGossipRunning -> context.json(isGossipRunning ? OK_STATUS : NOT_OK_STATUS))
+                         .onFailure(cause -> processFailure(cause, context, host, remoteAddress, request));
```
Comment (yifan-c) [inline server/src/main/java/org/apache/cassandra/sidecar/routes/GossipHealthHandler.java]: please align the dots. 
I thought it is already enforced by checkstyle, but not. I will make sure the rule is added.

File: server/src/test/java/org/apache/cassandra/sidecar/routes/CommonTest.java [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+
+    class CommonTestModule extends AbstractModule
+    {
+        @Provides
+        @Singleton
+        public InstancesMetadata instanceConfig()
+        {
+            final int instanceId = 100;
+            final String host = "127.0.0.1";
+            final InstanceMetadata instanceMetadata = mock(InstanceMetadata.class);
```
Comment (yifan-c) [inline server/src/test/java/org/apache/cassandra/sidecar/routes/CommonTest.java]: the `final` modifier in the method scope is usually omitted. See https://cassandra.apache.org/_/development/code_style.html#boilerplate

> never declare local variables and parameters final.

**Review by yifan-c — COMMENTED** [review]
Looks good to me. There are a few comments on codestyle. Please update. 

Besides that, please add an entry to CHANGES.txt

File: server/src/test/java/org/apache/cassandra/sidecar/routes/CommonTest.java (hunk 23 above)
Comment (skoppu22) [reply server/src/test/java/org/apache/cassandra/sidecar/routes/CommonTest.java]: This code is moved here from GossipInfoHandlerTest, don't know why were they marked final. Removed them

**Review by frankgh — APPROVED** [review]
+1

**Comment by bbotella** [comment]
+1 (nb)

**Review by bbotella — APPROVED** [review]

**Review by yifan-c — APPROVED** [review]
